File: python/241.py
"""
https://leetcode.com/problems/different-ways-to-add-parentheses/

Given a string of numbers and operators, return all possible results from computing all the different possible ways to group numbers and operators. The valid operators are +, - and *.

Example 1:

Input: "2-1-1"
Output: [0, 2]
Explanation:
((2-1)-1) = 0
(2-(1-1)) = 2
Example 2:

Input: "2*3-4*5"
Output: [-34, -14, -10, -10, 10]
Explanation:
(2*(3-(4*5))) = -34
((2*3)-(4*5)) = -14
((2*(3-4))*5) = -10
(2*((3-4)*5)) = -10
(((2*3)-4)*5) = 10
"""


from typing import List

# 这是一个非常好的题

class Solution:
    def diffWaysToCompute(self, input: str) -> List[int]:
        import re, operator
        tokens = re.split(r'(\D)', input)
        nums = [int(i) for i in tokens[::2]]
        ops = tokens[1::2]
        M = {'+': operator.add, '-': operator.sub, '*': operator.mul}

        def build(low, high):
            if low == high:
                return [nums[low]]
            res = []
            # ops nums之间有某种对应关系
            for i in range(low, high):
                a = build(low, i)
                b = build(i+1, high)
                #tmp = [M[ops[i]](p, q) for p in a for q in b]
                #print(a, ops[i], b, tmp)
                res.extend([M[ops[i]](p, q) for p in a for q in b])
            return res

        return build(0, len(nums) - 1)


if __name__ == '__main__':
    print(Solution().diffWaysToCompute("2*3-4*5"))

File: go/241.go
/*

https://leetcode.com/problems/different-ways-to-add-parentheses/

Given a string of numbers and operators, return all possible results from computing all the different possible ways to group numbers and operators. The valid operators are +, - and *.

Example 1:

Input: "2-1-1"
Output: [0, 2]
Explanation:
((2-1)-1) = 0
(2-(1-1)) = 2
Example 2:

Input: "2*3-4*5"
Output: [-34, -14, -10, -10, 10]
Explanation:
(2*(3-(4*5))) = -34
((2*3)-(4*5)) = -14
((2*(3-4))*5) = -10
(2*((3-4)*5)) = -10
(((2*3)-4)*5) = 10

*/


package main


import "fmt"


func parse(s string) []string {
    tokens := make([]string, 0)
    n := ""
    for i := 0; i < len(s); i++ {
        c := s[i]
        if c == ' ' {
            // pass
        } else if c == '+' || c == '-' || c == '*' {
            if len(n) > 0 {
                tokens = append(tokens, n)
                n = ""
            }
            tokens = append(tokens, string([]byte{c}))
        } else {
            n = n + string([]byte{c})
        }
    }
    if len(n) > 0 {
        tokens = append(tokens, n)
    }
    return tokens
}



func calculate(op string, a, b int) int {
    if op == "+"  {
        return a + b
    }
    if op == "-" {
        return a - b
    }
    return a * b
}


func diffWaysToCompute(input string) []int {
    vals := make([]int, 0)
    ops := make([]string, 0)
    tokens := parse(input)
    fmt.Println(tokens)
    for _, token := range tokens {
        if token == "+" || token == "-" ||  token == "*" {
            ops = append(ops, token)
        } else {
            t := 0
            for i := 0; i < len(token); i++ {
                t = t * 10 + int(token[i] - '0')
            }
            vals = append(vals, t)
        }
    }
    fmt.Println(ops, vals)

    // 本质上是构建一棵二叉排序树?
    // 为什么你想不到这个算法呢?
    // 或者是说你为什么建立不起来这个联系呢?
    var f func(left, right int) []int
    f = func(left, right int) []int {
        if left == right {
            return []int{vals[left]}
        } 
        res := make([]int, 0)
        for i := left; i < right; i++ {
            a := f(left, i)
            b := f(i+1, right)
            for _, p := range a {
                for _, q := range b {
                    c := calculate(ops[i], p, q)
                    res = append(res, c)
                }
            } 
        }
        return res
    }
    return f(0, len(vals) - 1) 
}


func main() {
    s := "2*3-4*5"
    fmt.Println(diffWaysToCompute(s))
}
